NOISSUE - Add Subscriber Config

rodneyosodo opened the pull request:
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

### What does this do?
Add subscriber config to differentiate between adapters and writers. This is because writers can consume old messages that are already in nats jetstream while adapters consume new messages only.

### Which issue(s) does this PR fix/relate to?
No issue

### List any changes that modify/break current functionality
Change subscribe and unsubscribe functions to 
```go
Subscribe(ctx context.Context, cfg SubscriberConfig) error

Unsubscribe(ctx context.Context, cfg SubscriberConfig) error
```

Subscriber config constitutes

```go
type SubscriberConfig struct {
	ID            string
	Topic         string
	Handler       MessageHandler
	DeliverPolicy DeliverPolicy
}
``` 
### Have you included tests for your changes?
No tests

### Did you document any new/modified functionality?
Not yet

### Notes
To be merged after https://github.com/mainflux/mainflux/pull/1893

rodneyosodo committed:
Replace Nats with Nats Jestream For PubSub

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Add Stream Description

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Setup subscriber config to contain handler topic and ID

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Add delivery policy

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

SammyOina commented on pkg/messaging/pubsub.go:
```diff
@@ -23,14 +34,21 @@ type MessageHandler interface {
+	ID            string
+	Topic         string
+	Handler       MessageHandler
+	DeliverPolicy DeliverPolicy
```
```suggestion
	DeliveryPolicy DeliveryPolicy
```

rodneyosodo committed:
Avoid duplicate messages

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Rename to DeliveryPolicy

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Fix tests

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Not check for data result set when we are returning subset of messages

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
Fix connection leak in NATS publisher

The publisher struct in pkg/messaging/nats/publisher.go was modified to include a new `conn` field of type `*broker.Conn`. This change was made to fix a connection leak issue in the NATS publisher.

The `NewPublisher` function was updated to assign the `conn` parameter to the new `conn` field in the publisher struct.

Additionally, the `Close` method in the publisher struct was modified to close the `conn` connection.

This commit fixes the connection leak issue in the NATS publisher and ensures that connections are properly closed.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

dborovcanin commented on pkg/messaging/pubsub.go:
```diff
@@ -5,6 +5,17 @@ package messaging
+	DeliverNewPolicy DeliveryPolicy = iota
+
+	// DeliverAllPolicy starts delivering messages from the very beginning of a stream.
+	DeliverAllPolicy
```
This oddly resembles [JetStream deliver policy ](https://docs.nats.io/nats-concepts/jetstream/consumers#deliverpolicy. That's fine, but we need to check if there are other policies we'd like to support, such as deliver by start time, which requires additional context to the policy. We also need to check how it resonates with other brokers. Let's accept this one, but please open a ticket for potential improvements.

rodneyosodo replied:
https://github.com/absmach/magistrala/issues/9

dborovcanin commented on pkg/messaging/pubsub.go:
```diff
@@ -23,14 +34,21 @@ type MessageHandler interface {
 	// Unsubscribe unsubscribes from the message stream and
 	// stops consuming messages.
-	Unsubscribe(ctx context.Context, id, topic string) error
+	Unsubscribe(ctx context.Context, cfg SubscriberConfig) error
```
Let's keep ID and topic for now, since `cfg` can be a little confusing in the context of `Unsubscribe` (and other fields are not used and probably won't be used in the future, either).

rodneyosodo committed:
For unsubscribe remove config

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

drasko commented on consumers/writers/mongodb/consumer.go:
```diff
@@ -45,7 +46,17 @@ func (repo *mongoRepo) saveSenml(ctx context.Context, messages interface{}) erro
 	var dbMsgs []interface{}
 	for _, msg := range msgs {
-		dbMsgs = append(dbMsgs, msg)
+		// check if message is already in database
```
Start comments with capital letter

rodneyosodo committed:
Fix comment

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

drasko reviewed (approved):
LGTM
